DataFormats/Common/interface/ThinnedRefSet.h

```diff
@@ -0,0 +1,50 @@
+#ifndef DataFormats_Common_ThinnedRefSet_h
+#define DataFormats_Common_ThinnedRefSet_h
+
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/Common/interface/Ref.h"
+#include "DataFormats/Common/interface/RefItemGet.h"
+
+#include <unordered_set>
+
+namespace edm {
+  class EDProductGetter;
+
+  template <typename C>
+  class ThinnedRefSet {
+  public:
+    class Filler {
+    public:
+      Filler(ThinnedRefSet<C>* set, RefProd<C> refProd, edm::EDProductGetter const& prodGetter)
+          : set_(set), refProd_(refProd), prodGetter_(prodGetter) {}
+
+      template <typename T, typename F>
+      void insert(Ref<C, T, F> const& ref) {
+        if (ref.isNull())
+          return;
+        auto thinnedRef = thinnedRefFrom(ref, refProd_, prodGetter_);
+        set_->keys_.insert(thinnedRef.key());
+      }
+
+    private:
+      ThinnedRefSet<C>* set_;
+      RefProd<C> refProd_;
+      edm::EDProductGetter const& prodGetter_;
+    };
+
+    ThinnedRefSet() = default;
+
+    Filler fill(RefProd<C> refProd, edm::EDProductGetter const& prodGetter) {
+      return Filler(this, refProd, prodGetter);
+    }
+
+    void clear() { keys_.clear(); }
+
+    bool contains(unsigned int key) const { return keys_.find(key) != keys_.end(); }
+
+  private:
+    std::unordered_set<unsigned int> keys_;
```

Comment:
This could be also `std::vector<bool>` with the size of the largest encountered key. There is some tradeoff on how sparse the chosen set of keys is, but I haven't done any analysis on that (probably should do).

Reply:
As a guess, I’d think a `std::vector<unsigned int>` that was reserved and kept sorted might be faster and more memory efficient as the internals of `unordered_set` uses `std::list` for items in same hash bucket.

Reply:
Thanks. Compared to `std::vector<unsigned int>`, `std::vector<bool>` would be smaller if more than 1/32 of the parent collection elements are included in the thinned collection. I'll anyway leave a possible change to a later PR.